test/middleware.js

```diff
@@ -49,6 +49,151 @@ describe('middleware', () => {
       .parse()
   })
 
+  it('runs the pre-check middlware before reaching the handler', function (done) {
+    yargs(['mw'])
+      .middleware(function (argv) {
+        argv.mw = 'mw'
+      }, true)
+      .command(
+        'mw',
+        'adds func to middleware',
+        {
+          'mw': {
+            'demand': true,
+            'string': true
+          }
+        },
+        function (argv) {
+          // we should get the argv filled with data from the middleware
+          argv.mw.should.equal('mw')
+          return done()
+        }
+      )
+      .exitProcess(false) // defaults to true.
+      .parse()
+  })
+
+  it('runs the pre-check middleware and ensures theres a context object with commands and availableOptions', function (done) {
+    yargs(['mw'])
+      .middleware(function (argv, context) {
+        argv.mw = context.commands[0]
+        argv.other = Array.isArray(context.availableOptions.mw)
+      }, true)
+      .command(
+        'mw',
+        'adds func to middleware',
+        {
+          'mw': {
+            'demand': true,
+            'string': true
+          }
+        },
+        function (argv) {
+          // we should get the argv filled with data from the middleware
+          argv.mw.should.equal('mw')
+          argv.other.should.equal(true)
+          return done()
+        }
+      )
+      .exitProcess(false) // defaults to true.
+      .parse()
+  })
+
+  it('runs the pre-check middlware with an array passed in and ensures theres a context object with commands and availableOptions', function (done) {
+    yargs(['mw'])
+      .middleware([function (argv, context) {
+        argv.mw = context.commands[0]
+        argv.other = Array.isArray(context.availableOptions.mw)
+      }], true)
+      .command(
+        'mw',
+        'adds func to middleware',
+        {
+          'mw': {
+            'demand': true,
+            'string': true
+          }
+        },
+        function (argv) {
+          // we should get the argv filled with data from the middleware
+          argv.mw.should.equal('mw')
+          argv.other.should.equal(true)
+          return done()
+        }
+      )
+      .exitProcess(false) // defaults to true.
+      .parse()
+  })
+
+  it('runs the pre-check middlware ensures if an async function is ran it throws an error', function (done) {
+    try {
+      yargs(['mw'])
+        .middleware([async function (argv, context) {
```

Comment:
tests are failing because Node@6 doesn't support the `async` keyword out of the box, I'd just make this a `new Promise` and you should go green 👍